skyportal/handlers/api/plan.py

```diff
@@ -0,0 +1,450 @@
+
+    return params
+
+
+def gen_structs(params):
+
+    print('Loading skymap')
+    # Function to read maps
+    map_struct = gwemopt.utils.read_skymap(
+        params, is3D=params["do3D"], map_struct=params['map_struct']
+    )
+    print('Generating MOC struct')
+    moc_structs = gwemopt.skyportal.create_moc_from_skyportal(
+        params, map_struct=map_struct
+    )
+    tile_structs = gwemopt.tiles.moc(params, map_struct, moc_structs)
+
+    if "previous_coverage_struct" in params:
+        tile_structs, coverage_struct = gwemopt.coverage.timeallocation(
+            params,
+            map_struct,
+            tile_structs,
+            previous_coverage_struct=params["previous_coverage_struct"],
+        )
+    else:
+        tile_structs, coverage_struct = gwemopt.coverage.timeallocation(
+            params, map_struct, tile_structs
+        )
+
+    if params["doPlots"]:
+        gwemopt.plotting.skymap(params, map_struct)
+        gwemopt.plotting.tiles(params, map_struct, tile_structs)
+        gwemopt.plotting.coverage(params, map_struct, coverage_struct)
+
+    return map_struct, tile_structs, coverage_struct
+
+
+def create_plan(
+    request_handler,
+    instrument,
+    localization,
+    validity_window_start=None,
+    validity_window_end=None,
+    plan_name=None,
+    **plan_args,
+):
+
+    dateobs = localization.dateobs
+
+    if validity_window_start is None:
+        validity_window_start = datetime.datetime.now()
+    if validity_window_end is None:
+        validity_window_end = validity_window_start + datetime.timedelta(1)
+
+    plan_args = dict(plan_args)
+    plan_args.setdefault(
+        'tobs',
+        [
+            time.Time(validity_window_start).mjd - time.Time(dateobs).mjd,
+            time.Time(validity_window_end).mjd - time.Time(dateobs).mjd,
+        ],
+    )
+
+    exposuretimes = plan_args['exposuretimes']
+    if plan_args['doDither'] and instrument.nickname == 'DECam':
+        # Add dithering
+        exposuretimes = [2 * x for x in exposuretimes]
+
+    plan_args.setdefault('probability', 0.9)
+
+    if plan_name is None:
+        plan_name = "%s_%s_%s_%d_%d_%s_%d_%d" % (
+            localization.localization_name,
+            "".join(plan_args['filt']),
+            plan_args['schedule_type'],
+            plan_args['doDither'],
+            plan_args['doReferences'],
+            plan_args['filterScheduleType'],
+            exposuretimes[0],
+            100 * plan_args['probability'],
+        )
+
+    try:
+        plan = ObservingPlan.query.filter_by(plan_name=plan_name).one()
+    except NoResultFound:
+        plan = ObservingPlan(
+            dateobs=localization.dateobs,
+            plan_name=plan_name,
+            instrument_id=instrument.id,
+            validity_window_start=validity_window_start,
+            validity_window_end=validity_window_end,
+            plan_args=plan_args,
+        )
+        DBSession().add(plan)
+        request_handler.verify_and_commit()
```

Comment:
possible to hold off until the very end before committing?

Reply:
We need to tell the code we started making the plan so it is PENDING.

Reply:
okay, yeah makes sense for polling purposes.